lib/elixir/lib/string.ex

```diff
@@ -618,6 +618,10 @@ defmodule String do
   @spec replace_leading(t, t, t) :: t
   def replace_leading(string, match, replacement)
       when is_binary(string) and is_binary(match) and is_binary(replacement) do
+    if match == "" do
```

Comment:
I wouldn't worry about overhead. The problem with your example is that then we say that the problem with `replace_leading(:not_a_binary, "", %{hello: "mom, look not a binary!"})` will be that the match is `""` instead of the other two arguments not being binaries. So the solution would be to add `is_binary/1` clauses to the `""` clause as well, but at that point I prefer the `if` version. ¯\\\_(ツ)\_/¯ I can change if other people prefer that, but personally I think the difference is not very strong anyways.